### PR TITLE
[t137043] Create bt_ci_on_promise.yml

### PR DESCRIPTION
**Pull request by BT-rmartin**
This action allows to generate a tar.gz file with the project code having some modules cythonized and others not. 
That file is in the end attached as an artifact of a release that is automatically created, taking care even from existing releases of the same day creating different names
image

Non obfuscated
 - Project code
 - Braintec Odoo Apps they purchased (BT-Store)
 - OCA or other third-party repos

Obfuscated
 - All bt repos (except (part of) BT Store)

This assumes that we can have three new keys in the deploy.yaml that can help to get the expected result here
 1. non_project_modules:
    - A YAML sequence of additional modules to be installed, beyond the project-specific modules.
      This helps in filtering out modules that we deliver from submodules.
 2. non_bt_repos_to_obfuscate:
    - A comma-separated string representing repository names (not starting with 'bt-') whose modules need obfuscation.
 3. bt_modules_not_to_obfuscate:
    - A comma-separated string indicating 'bt-' prefixed modules that shouldn't be obfuscated.
      Typically, these are modules from BT-Store for which the customer has a license.

The main goal is obfuscating (all) modules from bt-repositories which are submodules of the project repository
According to the configurations above, we can extend the list of repositories to be obfuscated, or we can avoid obfuscating some particular bt modules
We also won't deliver all modules from the bt-repositories but just those indicated in non_project_modules

In the resulting tar.gz file:
-  there are no git files
- bt submodules have been removed and their code is now in the folder bt-modules (obfuscated or not according to configuration)
- submodules under ext are not obfuscated
image

 
<div> Links to Odoo: </div> <ul>
<li><a target="_blank" href="https://braintec.com/web#view_type=form&model=project.task&id=137043">[t137043] Cythonizing all for external repos</a></li>
</ul>

### CODE REVIEW

File: .github/workflows/bt_ci_on_premise.yml [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                    echo "==========================================================================================="
+                    mv $folder ../../bt_modules/$submodule
+                # If the module is in the list of non project modules to install, it's obfuscated and moved to the folder bt_modules
+                elif [[ " ${NON_PROJECT_MODULES_ARRAY[@]} " =~ " ${folder} " ]]; then
+                    echo "==========================================================================================="
+                    echo "OBFUSCATING MODULE $folder"
+                    echo "==========================================================================================="
+                    ../../../bt-script/cythonizer/py2cy $folder --exclude 'test.*|-\w+\.py' -f
+                    mv build/$folder ../../bt_modules/$submodule
+                    rm -rf build
```
Comment (BT-sschmid) [inline .github/workflows/bt_ci_on_premise.yml:140]: why not install cythonizer in the setup step using `pip install bt-script/cythonizer`?
then you can call py2cy directly.

File: .github/workflows/bt_ci_on_premise.yml [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+      - name: Check for existing release and construct name
+        id: construct_release_name
+        run: |
+          DATE_STR=$(date +'%Y%m%d')
+          RELEASE_NAME="$REF_NAME.$DATE_STR"
+          COUNTER=0
+          
+          # Use GitHub API to check if release with name exists
+          while curl -s -H "Authorization: token ${{ secrets.GITHUB_TOKEN }}" "https://api.github.com/repos/${{ github.repository }}/releases/tags/$RELEASE_NAME" | jq -e .name > /dev/null; do
```
Comment (BT-sschmid) [inline .github/workflows/bt_ci_on_premise.yml:180]: here you could also just use github's cli client: https://cli.github.com/manual/gh_release_list
It is installed by default in gh-actions

File: .github/workflows/bt_ci_on_premise.yml [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        with:
+          artifacts: "./${{ env.REPO_NAME }}.tar.gz"
+          artifactContentType: application/gzip
+          commit: ${{ github.sha }}
+          draft: false
+          name: ${{ env.RELEASE_NAME }}
+          prerelease: false
+          tag: ${{ env.RELEASE_NAME }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          generateReleaseNotes: true
```
Comment (BT-sschmid) [inline .github/workflows/bt_ci_on_premise.yml:201]: also here. Also I'm not sure if we should trust external actions...
https://cli.github.com/manual/gh_release_create

**Review by BT-sschmid — APPROVED** [review]
Good job and very well documented. 
I just have some points which might optionally be improved below.

File: .github/workflows/bt_ci_on_premise.yml [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+              echo "  $module"
+          done
+
+          # 2. Extracting non-bt repos to obfuscate from deploy.yaml
+          non_bt_repos=$(grep "non_bt_repos_to_obfuscate:" deploy.yaml | sed 's/non_bt_repos_to_obfuscate: //')
+          IFS=',' read -ra NON_BT_REPOS_ARRAY <<< "$non_bt_repos"
+                    
+          # 3. Extracting bt-modules not to obfuscate from deploy.yaml
+          bt_modules_not_to_obfuscate=$(grep "bt_modules_not_to_obfuscate:" deploy.yaml | sed 's/bt_modules_not_to_obfuscate: //')
+          IFS=',' read -ra BT_MODULES_NOT_TO_OBFUSCATE_ARRAY <<< "$bt_modules_not_to_obfuscate"
```
[the remark this answers is not in the document]
Comment (BT-rmartin) [reply .github/workflows/bt_ci_on_premise.yml:75]: Ok I will change it, for both

File: .github/workflows/bt_ci_on_premise.yml (hunk 1 above)
Comment (BT-rmartin) [reply .github/workflows/bt_ci_on_premise.yml:140]: Good input, I had several issues with the relative path, so perfect

File: .github/workflows/bt_ci_on_premise.yml (hunk 2 above)
Comment (BT-rmartin) [reply .github/workflows/bt_ci_on_premise.yml:180]: Ok, with gh release view we can also get it

File: .github/workflows/bt_ci_on_premise.yml (hunk 3 above)
Comment (BT-rmartin) [reply .github/workflows/bt_ci_on_premise.yml:201]: Ok, I agree, let's see if it works

**Review by BT-sschmid — APPROVED** [review]

**Pull request merged**